src/main/java/bio/terra/service/snapshot/flight/delete/SnapshotDeleteFlight.java

```diff
@@ -33,15 +38,31 @@ public SnapshotDeleteFlight(FlightMap inputParameters, Object applicationContext
         BigQueryPdao bigQueryPdao = (BigQueryPdao)appContext.getBean("bigQueryPdao");
         ResourceService resourceService = (ResourceService) appContext.getBean("resourceService");
         IamService iamClient = (IamService)appContext.getBean("iamService");
+        DatasetDao datasetDao = (DatasetDao) appContext.getBean("datasetDao");
         DatasetService datasetService = (DatasetService)appContext.getBean("datasetService");
         ConfigurationService configService = (ConfigurationService)appContext.getBean("configurationService");
 
         UUID snapshotId = UUID.fromString(inputParameters.get(
             JobMapKeys.SNAPSHOT_ID.getKeyName(), String.class));
         AuthenticatedUserRequest userReq = inputParameters.get(
             JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+        UUID datasetId;
 
         addStep(new LockSnapshotStep(snapshotDao, snapshotId, true));
+
+        // Lock the source dataset while deleting ACLs to avoid a race condition
+        // TODO note that with multi-dataset snapshots this will need to change
+        List<UUID> sourceDatasetIds;
+        try {
+            sourceDatasetIds = snapshotService.getSourceDatasetIdsFromSnapshotId(snapshotId);
+            datasetId = sourceDatasetIds.get(0);
+        } catch (SnapshotNotFoundException notFoundEx) {
+            datasetId = null;
+        }
+        if (datasetId != null) {
+            addStep(new LockDatasetStep(datasetDao, datasetId, false));
```

Comment:
Oh I didn't realize that!

Reply:
Hmm from here it actually looks like it would be shared with other flights: https://github.com/DataBiosphere/jade-data-repo/blob/17ca15357c81bbf38a98120e3c5300e7b725e1d0/src/main/java/bio/terra/service/dataset/DatasetDao.java#L151

Reply:
After looking at the code I'm not sure what version of locking we want to use here. Maybe there's another jade person who knows more about this part of the code.

My concern is that all current uses of `LockDatasetStep` use `sharedLock = true`. So I'm not sure if we need to use `false` here. Also I see that other uses of locking also use a retry to avoid failing the step if the lock can't be acquired.